File: web/backend/src/db/models/tool.py
"""Tool ORM model for storing tool records (builtin and third-party)."""
import uuid
from datetime import datetime

from sqlalchemy import JSON, DateTime, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from db.base import Base


class Tool(Base):
    """Tool record — tracks name, category, source, status, and parameter definitions."""

    __tablename__ = "tools"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(128), unique=True, index=True, comment="工具标识符，如 execute_code")
    display_name: Mapped[str] = mapped_column(String(128), comment="显示名称，如 代码执行")
    description: Mapped[str] = mapped_column(Text, default="", comment="工具描述")
    category: Mapped[str] = mapped_column(
        String(32), default="other", comment="功能分类: code/network/message/file/data/ai/system/other"
    )
    source: Mapped[str] = mapped_column(
        String(32), default="builtin", comment="来源: builtin/third_party"
    )
    status: Mapped[str] = mapped_column(
        String(32), default="enabled", comment="状态: enabled/disabled/unavailable"
    )
    version: Mapped[str] = mapped_column(String(32), default="1.0.0", comment="版本号")
    author: Mapped[str] = mapped_column(String(128), default="", comment="作者")
    tags: Mapped[list] = mapped_column(JSON, default=list, comment="标签数组")
    params: Mapped[list] = mapped_column(JSON, default=list, comment="参数定义数组 [{key, label, required, type}]")

    # ── 改进1：DB 驱动工具注册表 ──────────────────────────────────────
    # implementation: 内置函数工具存储 module_path:func_name，如 agent.tools.http_request:http_request
    #                 MCP 工具存储 MCP 目录中的 name（对应 mcp_tools.name）
    implementation: Mapped[str | None] = mapped_column(
        String(256), nullable=True,
        comment="工具实现路径: function 类型为 module_path:func_name; mcp 类型为 mcp_tool_name"
    )
    # tool_type: 区分内置函数工具、MCP 工具、第三方插件
    tool_type: Mapped[str] = mapped_column(
        String(32), default="function",
        comment="工具类型: function(内置函数) | mcp(MCP工具) | plugin(第三方插件)"
    )

    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), comment="创建时间")
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

File: web/frontend/src/types/tool.ts
/** 工具状态 */
export type ToolStatus = 'enabled' | 'disabled' | 'unavailable'

/** 工具来源 */
export type ToolSource = 'builtin' | 'third-party'

/** 工具功能分类 */
export type ToolCategory =
  | 'code'
  | 'network'
  | 'message'
  | 'file'
  | 'data'
  | 'ai'
  | 'system'
  | 'other'

/** 工具参数定义 */
export interface ToolParam {
  key: string
  label: string
  required: boolean
  type: string
}

/** 工具实体 */
export interface Tool {
  id: string
  name: string
  displayName: string
  description: string
  category: ToolCategory
  source: ToolSource
  status: ToolStatus
  version: string
  author: string
  usedByAgents: string[]
  tags: string[]
  params?: ToolParam[]
  createdAt: string
  updatedAt: string
}

/** 创建/更新工具请求 */
export interface ToolCreateRequest {
  name: string
  displayName: string
  description?: string
  category?: ToolCategory
  status?: ToolStatus
  version?: string
  tags?: string[]
  params?: ToolParam[]
}

/** 分类元数据 */
export interface CategoryMeta {
  label: string
  icon: string
  color: string
  bg: string
  border: string
}

/** 功能分类映射 */
export const CATEGORY_META: Record<ToolCategory, CategoryMeta> = {
  code: { label: '编码工具', icon: 'Code2', color: 'var(--color-tool-blue)', bg: 'rgba(59,130,246,0.08)', border: 'rgba(59,130,246,0.2)' },
  network: { label: '网络工具', icon: 'Globe', color: 'var(--color-tool-cyan)', bg: 'rgba(6,182,212,0.08)', border: 'rgba(6,182,212,0.2)' },
  message: { label: '消息工具', icon: 'MessageSquare', color: 'var(--color-tool-green)', bg: 'rgba(34,197,94,0.08)', border: 'rgba(34,197,94,0.2)' },
  file: { label: '文件工具', icon: 'FileText', color: 'var(--color-tool-amber)', bg: 'rgba(245,158,11,0.08)', border: 'rgba(245,158,11,0.2)' },
  data: { label: '数据工具', icon: 'Database', color: 'var(--color-tool-purple)', bg: 'rgba(168,85,247,0.08)', border: 'rgba(168,85,247,0.2)' },
  ai: { label: 'AI 工具', icon: 'Sparkles', color: 'var(--color-tool-pink)', bg: 'rgba(236,72,153,0.08)', border: 'rgba(236,72,153,0.2)' },
  system: { label: '系统工具', icon: 'Cpu', color: 'var(--color-tool-indigo)', bg: 'rgba(99,102,241,0.08)', border: 'rgba(99,102,241,0.2)' },
  other: { label: '其他', icon: 'Wrench', color: 'var(--color-tool-gray)', bg: 'rgba(148,163,184,0.08)', border: 'rgba(148,163,184,0.2)' },
}

/** 状态元数据 */
export interface StatusMeta {
  label: string
  color: string
  bg: string
  border: string
  icon: string
}

export const STATUS_META: Record<ToolStatus, StatusMeta> = {
  enabled: { label: '已启用', color: 'var(--color-tool-emerald)', bg: 'rgba(16,185,129,0.08)', border: 'rgba(16,185,129,0.2)', icon: 'CheckCircle2' },
  disabled: { label: '已禁用', color: 'var(--color-tool-amber)', bg: 'rgba(245,158,11,0.08)', border: 'rgba(245,158,11,0.2)', icon: 'PauseCircle' },
  unavailable: { label: '不可用', color: 'var(--color-tool-gray)', bg: 'rgba(148,163,184,0.06)', border: 'rgba(148,163,184,0.15)', icon: 'AlertCircle' },
}

/** 来源标签 */
export const SOURCE_LABELS: Record<ToolSource, string> = {
  builtin: '内置',
  'third-party': '第三方',
}

/** 工具列表响应 */
export interface ToolListResponse {
  items: Tool[]
  total: number
  page: number
  page_size: number
}
